### PR TITLE
Some shuttes rework. Allow many shuttes of one template to work at the same time, and returned 2 fighter to syndicate ambush.

### PR DESCRIPTION
**Pull request by Dennok**
## About The Pull Request
Make docking_port gen id and name.
This allow creating and working of many shuttles created by single template.

Add to shuttle initialisation behaveor to auto modification doors, cameras, camera consoles, buttons, pod doors id's to unicue per shuttle. Now Buttons from one shuttle dont open doors from another. And you cant see cameras from another copies of the ship.

Now ID tru unical, replaced part of it behaveor by destynation_type. You can set many docks of one destynation_type but all it have different id's. you can send shuttle to docks ID, Or can send shuttle to destynation_type, then it be send to first port with matches destynation_type.
Now shuttle console collect list of ID's by matching console(shuttle) and dock destynation_type. And you can send shutle to dock ID.

Previous console collect list of ID's by matching console(shuttle) and dock ID. And if list have matching ID's you you can send only to first match ID in SSshuttle.shuttles list.

And if ports have match name or id it can be dont show in some menu.

SSshuttle stationary and mobile now associated lists(id = shuttle/dock).
## Why It's Good For The Game
Now we dont need code new areas, consoles, docking ports to create new shuttles from single template.
If you need second whiteship, just place it.
If you want new docking port just place it, no need to VV it id, and VV shuttle console destinations. Just make sure that it destynation_type same, and it fit in width and height.
For whiteship's Deep Space just place whiteship dock template.

## Changelog
:cl:
add: Make many single template shuttles work at the same time.
add: Returned 2 fighter to syndicate ambush.
/:cl:

TODO:

- [x] Make `obj/docking_port` id unical, replace part of it behavior by `destination_type`
- [x] Replace all starting 'id' to `destination_type` in `obj/docking_port`
- [x] Make every shuttle have its own button codes, and cameras.
- [x] Change using shuttles.len in id to something another, to dont get messeg after deliting shuttles/docks.
- [x] Make map_template_load() propertly differentise machinery codes when load shuttles.
- [x] Make SSshuttle.mobile and SSshuttle.stationary asociative list's.
- [x] Sure that all shuttle related mechanics (what i find and test) work as intended.
- [ ] Tests
- [ ] Sure that shuttle related mechanics work as intended.
- [ ] Live tests
- [ ] Sure that absolutely all shuttle related mechanics work as intended.
- [ ] Change `destination_type` var name to some more intuitive and universal.
- [ ] Give shuttle manipulator variant to load addition shuttles, instead of replacing.
- [ ] Make shuttle manipulator less laggy.

### CODE REVIEW

**Comment by ninjanomnom** [comment]
The issue with multiple shuttles with one template isn't the docking port, the issue is with all shuttles made from one template sharing areas. We need changes in map loading in order to support that

File: code/modules/shuttle/shuttle.dm [hunk 1]
```diff
@@ -163,10 +165,8 @@
 /obj/docking_port/stationary/Initialize(mapload)
 	. = ..()
 	SSshuttle.stationary += src
-	if(!id)
-		id = "[SSshuttle.stationary.len]"
-	if(name == "dock")
-		name = "dock[SSshuttle.stationary.len]"
+	id = "[SSshuttle.stationary.len]"
+	name = "[name] dock[SSshuttle.stationary.len]"
```
Comment (ninjanomnom) [inline code/modules/shuttle/shuttle.dm]: Do not remove the previous code, this is to handle map edits

File: code/modules/shuttle/shuttle.dm [hunk 2]
```diff
@@ -290,10 +290,8 @@
 /obj/docking_port/mobile/Initialize(mapload)
 	. = ..()
 
-	if(!id)
-		id = "[SSshuttle.mobile.len]"
-	if(name == "shuttle")
-		name = "shuttle[SSshuttle.mobile.len]"
+	id = "[SSshuttle.mobile.len]"
+	name = "[name] shuttle[SSshuttle.mobile.len]"
```
Comment (ninjanomnom) [inline code/modules/shuttle/shuttle.dm]: Ditto

File: code/modules/shuttle/shuttle.dm [hunk 3]
```diff
@@ -315,10 +313,8 @@
 	var/list/static/shuttle_id = list()
 	var/idnum = ++shuttle_id[template]
 	if(idnum > 1)
-		if(id == initial(id))
-			id = "[id][idnum]"
-		if(name == initial(name))
-			name = "[name] [idnum]"
+		id = "[id][idnum]"
+		name = "[name] [idnum]"
```
Comment (ninjanomnom) [inline code/modules/shuttle/shuttle.dm]: Ditto

File: code/modules/shuttle/shuttle.dm [hunk 4]
```diff
@@ -432,7 +428,10 @@
 		if(initiate_docking(S1) != DOCKING_SUCCESS)
 			WARNING("shuttle \"[id]\" could not enter transit space. Docked at [S0 ? S0.id : "null"]. Transit dock [S1 ? S1.id : "null"].")
 		else
-			previous = S0
+			if(S0.delete_after)
+				S0.Destroy(TRUE)
```
Comment (ninjanomnom) [inline code/modules/shuttle/shuttle.dm]:
```suggestion
				qdel(S0, TRUE)
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by Dennok** [comment]
Templates actually creates different zones for spawned templates. And as i see it dont shares apc.
scrnshot1
Here i fix issue what from spawned from one template shuttles work only one. 
And its problem in sharing its shuttle IDs.

File: code/modules/shuttle/shuttle.dm (hunk 1 above)
Comment (Dennok) [reply code/modules/shuttle/shuttle.dm]: how to handle !00% different ID and NAME near shuttles in another way?

File: code/modules/shuttle/shuttle.dm (hunk 2 above)
Comment (Dennok) [reply code/modules/shuttle/shuttle.dm]: how to handle !00% different ID and NAME near shuttles in another way?

File: code/modules/shuttle/shuttle.dm (hunk 1 above)
Comment (Dennok) [reply code/modules/shuttle/shuttle.dm]: this code want change id if we have 2 or more of dock/shuttles but i dont see it work. so i remove the check and give unical suffix

**Comment by Dennok** [comment]
About ID and name
Actual: 
image
image

After my changes:
image

File: code/modules/shuttle/shuttle.dm (hunk 1 above)
Comment (ninjanomnom) [reply code/modules/shuttle/shuttle.dm]: I don't understand what you're saying here

Comment (Dennok) [reply code/modules/shuttle/shuttle.dm]: Problem in map edited shuttles, it ID and NAME same for all created by template.

Comment (ninjanomnom) [reply code/modules/shuttle/shuttle.dm]: Then remove those map edits causing problems and being used improperly. That's not a reason to remove map edit support

**Comment by ninjanomnom** [comment]
I'll have to check sometime but unless someone has made changes to map loading areas are shared by shuttles for sure. Shuttle movement as well doesn't account for handling existing area instances.

**Comment by Dennok** [comment]
I think this make shuttle areas dont unique.
https://github.com/tgstation/tgstation/blob/master/code/game/area/areas/shuttles.dm#L14

**Comment by Dennok** [comment]
supply sells, i belive.

File: code/modules/shuttle/shuttle.dm (hunk 1 above)
Comment (Dennok) [reply code/modules/shuttle/shuttle.dm]: I return map edit support

**Comment by SpaceManiac** [comment]
@ninjanomnom #39890

**Comment by ninjanomnom** [comment]
Oh yeah I forgot that was a thing

**Comment by Dennok** [comment]
Split it on parts?

**Comment by SpaceManiac** [comment]
I'm not sure splitting is needed. However, if it was, how would you split it? Mainly I am busy.

**Comment by Dennok** [comment]
I can split out part of code that changes door, poddoor. buttons and etc. codes, and 2 fighter. 
But it not to much.

**Comment by optimumtact** [comment]
@ninjanomnom also need feedback from you on final state

**Comment by Dennok** [comment]
To hard, and dirty. I separate it, and some rework.

**Pull request closed without merging**